src/lib/protocols/cassandra.c

```diff
@@ -0,0 +1,139 @@
+ * nDPI is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * nDPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with nDPI.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include "ndpi_protocol_ids.h"
+#include "ndpi_api.h"
+
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_CASSANDRA
+#define CASSANDRA_HEADER_LEN 9
+#define CASSANDRA_MAX_BODY_SIZE 268435456 //256MB (256 * 1024^2)
+
+enum cassandra_version
+{
+  CASSANDRA_V1_REQUEST = 0x01,
+  CASSANDRA_V1_RESPONSE = 0x81,
+  CASSANDRA_V2_REQUEST = 0x02,
+  CASSANDRA_V2_RESPONSE = 0x82,
+  CASSANDRA_V3_REQUEST = 0x03,
+  CASSANDRA_V3_RESPONSE = 0x83,
+  CASSANDRA_V4_REQUEST = 0x04,
+  CASSANDRA_V4_RESPONSE = 0x84
+};
+
+enum cassandra_opcode
+{
+  ERROR = 0x00,
+  STARTUP = 0x01,
+  READY = 0x02,
+  AUTHENTICATE = 0x03,
+  OPTIONS = 0x05,
+  SUPPORTED = 0x06,
+  QUERY = 0x07,
+  RESULT = 0x08,
+  PREPARE = 0x09,
+  EXECUTE = 0x0A,
+  REGISTER = 0x0B,
+  EVENT = 0x0C,
+  BATCH = 0x0D,
+  AUTH_CHALLENGE = 0x0E,
+  AUTH_RESPONSE = 0x0F,
+  AUTH_SUCCESS = 0x10
+};
+
+static bool ndpi_check_valid_cassandra_version(uint8_t version)
+{
+  switch(version) {
+    case CASSANDRA_V1_REQUEST:
+    case CASSANDRA_V1_RESPONSE:
+    case CASSANDRA_V2_REQUEST:
+    case CASSANDRA_V2_RESPONSE:
+    case CASSANDRA_V3_REQUEST:
+    case CASSANDRA_V3_RESPONSE:
+    case CASSANDRA_V4_REQUEST:
+    case CASSANDRA_V4_RESPONSE:
+      return true;
+  }
+  return false;
+}
+
+static bool ndpi_check_valid_cassandra_opcode(uint8_t opcode)
+{
+  switch (opcode) {
+    case ERROR:
+    case STARTUP:
+    case READY:
+    case AUTHENTICATE:
+    case OPTIONS:
+    case SUPPORTED:
+    case QUERY:
+    case RESULT:
+    case PREPARE:
+    case EXECUTE:
+    case REGISTER:
+    case EVENT:
+    case BATCH:
+    case AUTH_CHALLENGE:
+    case AUTH_RESPONSE:
+    case AUTH_SUCCESS:
+      return true;
+  }
+  return false;
+}
+
+void ndpi_search_cassandra(struct ndpi_detection_module_struct *ndpi_struct,
+                           struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct *packet = &flow->packet;
+
+  if (packet->tcp) {
+    if (packet->payload_packet_len >= CASSANDRA_HEADER_LEN &&
+        ndpi_check_valid_cassandra_version(get_u_int8_t(packet->payload, 0)) &&
+        ndpi_check_valid_cassandra_opcode(get_u_int8_t(packet->payload, 4)) &&
+        get_u_int32_t(packet->payload, 5) <= CASSANDRA_MAX_BODY_SIZE &&
+        get_u_int32_t(packet->payload, 5) >= (packet->payload_packet_len - CASSANDRA_HEADER_LEN)) {
```

Comment:
```suggestion
        get_int32_t(packet->payload, 5) >= (packet->payload_packet_len - CASSANDRA_HEADER_LEN)) {
```
may fix the CI issue and should be fine as `CASSANDRA_MAX_BODY_SIZE` < (INT_MAX /2)

Reply:
I tested locally and the function `get_int32_t()` doesn't seem to exist, so I fixed the warning by simply casting the right hand operand to `u_int32t`, as it should be always a valid operation

```
        get_u_int32_t(packet->payload, 5) >= (uint32_t) (packet->payload_packet_len - CASSANDRA_HEADER_LEN)) {
```

Reply:
Oh sorry. I thought here was a function. But an explicit cast seems also fine to me althought it should be used as less as possible.